File: go/oop-patterns/behavioral-patterns/memento/memento.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

// объект, состояние которого нужно сохранять и восстанавливать
// должен иметь методы для сохранения своего состояния в объект снимка и
// восстановления своего состояния из такого объекта
type originator struct {
	state string
}

func (o *originator) doSomething() {
	fmt.Println("Источник: Делаю что-то важное.")
	o.state = o.generateRandomString(30)
	fmt.Printf("Источник: моё состояние изменилось на:  %v\n", o.state)
}

func (o *originator) generateRandomString(length int) string {
	charSet := "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
	randChars := make([]rune, length)
	randNums := rand.Perm(len(charSet))
	for i := range randChars {
		randChars[i] = rune(charSet[randNums[i]])
	}
	return string(randChars)
}

func (o *originator) save() memento {
	return &concreteMemento{state: o.state, date: time.Now().Format(time.RFC822)}
}

func (o *originator) restore(memento memento) {
	o.state = memento.getState()
	fmt.Printf("Источник: моё состояние изменилось на:  %v\n", o.state)
}

// Интерфейс снимка должен давать возможность получить данные о состоянии
// обратно
type memento interface {
	getState() string

	getName() string

	getDate() string
}

type concreteMemento struct {
	state string
	date  string
}

func (m *concreteMemento) getState() string {
	return m.state
}

func (m *concreteMemento) getName() string {

	return fmt.Sprintf("%v / (%v...)", m.date, m.state[:10])
}

func (m *concreteMemento) getDate() string {
	return m.date
}

// Смотритель напрямую не зависит от конкретного снимка
// он так же не зависит от конкретного объекта, чьё состояние сохраняется
// его задача - хранить снимки и вызывать соответствующие методы
// других объектов для сохранения и восстановления состояния
type caretaker struct {
	mementos []memento

	originator *originator
}

func (c *caretaker) backup() {
	fmt.Println("\nСмотритель: Сохраняю состояние источника...")
	c.mementos = append(c.mementos, c.originator.save())
}

func (c *caretaker) undo() {
	length := len(c.mementos)
	if length < 1 {
		return
	}
	memento := c.mementos[length-1]
	c.mementos = c.mementos[:length-1]

	fmt.Printf("Смотритель: восстанавливаю состояние источника до: %v\n", memento.getName())
	c.originator.restore(memento)
}

func (c *caretaker) showHistory() {
	fmt.Println("Смотритель: вот весь текущий список объектов-снимков:")
	for _, memento := range c.mementos {
		fmt.Println(memento.getName())
	}
}

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// ------------------------------- Клиентский код -------------------------------
// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
func main() {
	originator := &originator{state: "Изначальное состояние"}
	caretaker := caretaker{originator: originator}

	caretaker.backup()
	originator.doSomething()

	caretaker.backup()
	originator.doSomething()

	caretaker.backup()
	originator.doSomething()

	fmt.Println("")
	caretaker.showHistory()

	fmt.Println("\nКлиент: откатываю состояние")
	caretaker.undo()

	fmt.Println("\nКлиент: откатываю ещё раз")
	caretaker.undo()
}

File: javascript/oop-patterns/behavioral-patterns/memento/memento.js
// объект, состояние которого нужно сохранять и восстанавливать
// должен иметь методы для сохранения своего состояния в объект снимка и
// восстановления своего состояния из такого объекта
class Originator {
  state;
  constructor(state) {
    this.state = state;
    console.log(`Источник: Моё изначальное состояние: ${state}`);
  }
  doSomething() {
    console.log('Источник: Делаю что-то важное.');
    this.state = this.generateRandomString(30);
    console.log(`Источник: моё состояние изменилось на: ${this.state}`);
  }

  generateRandomString(length = 10) {
    const charSet = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';

    return new Array(length)
      .fill('a')
      .map(() => charSet.charAt(Math.floor(Math.random() * charSet.length)))
      .join('');
  }

  save() {
    return new ConcreteMemento(this.state);
  }

  restore(memento) {
    this.state = memento.getState();
    console.log(`Источник: моё состояние изменилось на: ${this.state}`);
  }
}

class ConcreteMemento {
  state;
  date;

  constructor(state) {
    this.state = state;
    this.date = new Date().toISOString().slice(0, 19).replace('T', ' ');
  }

  getState() {
    return this.state;
  }

  getName() {
    return `${this.date} / (${this.state.substring(0, 9)}...)`;
  }

  getDate() {
    return this.date;
  }
}

// Смотритель напрямую не зависит от конкретного снимка
// он так же не зависит от конкретного объекта, чьё состояние сохраняется
// его задача - хранить объекты снимков и вызывать соответствующие методы
// других объектов для сохранения и восстановления состояния
class Caretaker {
  mementos = [];

  originator;

  constructor(originator) {
    this.originator = originator;
  }

  backup() {
    console.log('\nСмотритель: Сохраняю состояние источника...');
    this.mementos.push(this.originator.save());
  }

  undo() {
    if (!this.mementos.length) {
      return;
    }
    const memento = this.mementos.pop();

    if (memento) {
      console.log(
        `Смотритель: восстанавливаю состояние источника до: ${memento.getName()}`
      );
      this.originator.restore(memento);
    } else {
      throw new Error(
        'Отсутствуют объекты снимков для восстановления состояния'
      );
    }
  }

  showHistory() {
    console.log('Смотритель: вот весь текущий список объектов-снимков:');
    for (const memento of this.mementos) {
      console.log(memento.getName());
    }
  }
}

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
//------------------------------- Клиентский код -------------------------------
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
const originator = new Originator('Изначальное состояние');
const caretaker = new Caretaker(originator);

caretaker.backup();
originator.doSomething();

caretaker.backup();
originator.doSomething();

caretaker.backup();
originator.doSomething();

console.log('');
caretaker.showHistory();

console.log('\nКлиент: откатываю состояние\n');
caretaker.undo();

console.log('\nКлиент: откатываю ещё раз\n');
caretaker.undo();
